messages/src/lib.rs

```diff
@@ -1540,6 +1540,130 @@ impl Decode for Report {
     }
 }
 
+/// Represents a request to upload reports to DAP aggregators.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct UploadRequest {
+    reports: Vec<Report>,
+}
+
+impl UploadRequest {
+    /// The media type associated with this protocol message.
+    pub const MEDIA_TYPE: &'static str = "application/dap-upload-req";
+
+    pub fn new(reports: &[Report]) -> Self {
+        Self {
+            reports: reports.to_vec(),
+        }
+    }
+
+    pub fn reports(&self) -> &[Report] {
+        &self.reports
+    }
+}
+
+impl Encode for UploadRequest {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        encode_fixlen_items(bytes, &self.reports)
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        let mut length = 0;
+        for report in &self.reports {
+            length += report.encoded_len()?;
+        }
+
+        Some(length)
+    }
+}
+
+impl ParameterizedDecode<usize> for UploadRequest {
+    fn decode_with_param(length: &usize, bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let reports = decode_fixlen_items(*length, &(), bytes)?;
+
+        Ok(Self { reports })
+    }
+}
+
+/// Represents the status of a report upload.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ReportUploadStatus {
+    id: ReportId,
+    error: ReportError,
+}
+
+impl ReportUploadStatus {
+    pub fn new(id: ReportId, error: ReportError) -> Self {
+        Self { id, error }
+    }
+
+    pub fn report_id(&self) -> ReportId {
+        self.id
+    }
+
+    pub fn error(&self) -> ReportError {
+        self.error
+    }
+}
+
+impl Encode for ReportUploadStatus {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.id.encode(bytes)?;
+        self.error.encode(bytes)
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.id.encoded_len()? + self.error.encoded_len()?)
+    }
+}
+
+impl Decode for ReportUploadStatus {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let id = ReportId::decode(bytes)?;
+        let error = ReportError::decode(bytes)?;
+
+        Ok(Self { id, error })
+    }
+}
+
+/// Represents a response to a request to upload reports to DAP aggregators.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct UploadResponse {
+    status: Vec<ReportUploadStatus>,
+}
+
+impl UploadResponse {
+    /// The media type associated with this protocol message.
+    pub const MEDIA_TYPE: &'static str = "application/dap-upload-resp";
+
+    pub fn new(statuses: &[ReportUploadStatus]) -> Self {
+        Self {
+            status: statuses.to_vec(),
+        }
+    }
+
+    pub fn status(&self) -> &[ReportUploadStatus] {
+        &self.status
+    }
+}
+
+impl Encode for UploadResponse {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        encode_fixlen_items(bytes, &self.status)
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(ReportId::LEN + 1)
+    }
```

Comment:
Just as soon as I `push` I see this obvious bug. Pretty sure that the response isn't constant size anymore, oops....